Updated MapboxAudioGuidanceButton, MapboxCameraModeButton and MapboxExtendableButton to support square, oval and circle shapes.

tomaszrybakiewicz opened the pull request:
Closes [#1807](https://github.com/mapbox/navigation-sdks/issues/1807)

### Description

- Added Buttons Demo Activity to QA Test App.
- Refactored `MapboxAudioGuidanceButton`, `MapboxCameraModeButton`, `MapboxExtendableButton` to support different button shapes.
- Introduced followings styles that implement various button shapes
- Moved `MapboxExtendableButton` to `libnavui-base` module.
- Moved `MapboxCameraModeButton` to `livnavui-maps` module.
- Added `libnavui-resources`,  `AppCompat` and `ConstraintLayout` as implementation depedency to libnavui-base.
- Updated Metalava API files for libnavui-base, libnavui-dropin and libnavui-maps modules.

_**MapboxAudioGuidanceButton styles**_
| Style name | Example Preview |
|:---|:---:|
| MapboxStyleAudioGuidanceButton <br/>_(default)_ | <img height="52" alt="MapboxStyleAudioGuidanceButton" src="https://user-images.githubusercontent.com/2678039/175368128-9a91d9b1-d61a-476c-8bc7-efac56b0c7a7.png" /> |
| MapboxStyleAudioGuidanceButton.Oval | <img height="52" alt="MapboxStyleAudioGuidanceButton Oval" src="https://user-images.githubusercontent.com/2678039/175366793-3d1cb4e0-094d-4e45-b07d-c2c410de928d.png"/> |
| MapboxStyleAudioGuidanceButton.Square | <img height="52" alt="MapboxStyleAudioGuidanceButton Square" src="https://user-images.githubusercontent.com/2678039/175366796-137475e4-c235-47b1-ae08-596abc35b2ed.png"/> |
| MapboxStyleAudioGuidanceButton.Circle | <img height="52" alt="MapboxStyleAudioGuidanceButton Circle" src="https://user-images.githubusercontent.com/2678039/175366790-09402eee-fbd3-4bb8-8918-8bf4d796e4be.png"/> |

_**MapboxCameraModeButton styles**_
| Style name | Example Preview |
|:---|:---:|
| MapboxStyleCameraModeButton <br/>_(default)_ | <img height="52" alt="MapboxStyleCameraModeButton" src="https://user-images.githubusercontent.com/2678039/175366803-1148d0b4-3cde-4f79-bafc-f3228a6e55e8.png" /> |
| MapboxStyleCameraModeButton.Oval | <img height="52" alt="MapboxStyleCameraModeButton Oval" src="https://user-images.githubusercontent.com/2678039/175366802-9543c711-1449-417a-ad6a-14454c261035.png"/> |
| MapboxStyleCameraModeButton.Square | <img height="52" alt="MapboxStyleCameraModeButton Square" src="https://user-images.githubusercontent.com/2678039/175366804-b78a1718-57a1-4091-b722-37d2eaeb5c78.png"/> |
| MapboxStyleCameraModeButton.Circle | <img height="52" alt="MapboxStyleCameraModeButton Circle" src="https://user-images.githubusercontent.com/2678039/175366800-e6a6eed6-9f70-492a-995f-90a0f27314cf.png" /> |

_**MapboxExtendableButton styles**_
| Style name | Example Preview |
|:---|:---:|
| MapboxStyleExtendableButton <br/>_(default)_ | <img height="52" alt="MapboxStyleExtendableButton" src="https://user-images.githubusercontent.com/2678039/175366808-578e47a2-4438-4f66-81f7-cb7fc96ce07c.png" /> |
| MapboxStyleExtendableButton.Oval | <img height="52" alt="MapboxStyleExtendableButton Oval" src="https://user-images.githubusercontent.com/2678039/175366807-78139e6e-2349-4411-8e45-d222b650d7f8.png" /> |
| MapboxStyleExtendableButton.Square | <img height="52" alt="MapboxStyleExtendableButton Square" src="https://user-images.githubusercontent.com/2678039/175366811-2b205113-0805-42e1-8531-b9115c0fc985.png"/> |
| MapboxStyleExtendableButton.Circle |  <img height="52" alt="MapboxStyleExtendableButton Circle" src="https://user-images.githubusercontent.com/2678039/175366805-26ef9750-a1da-4496-963c-98c1eeadf489.png" /> |

### Screenshots or Gifs

_(screen recording of qa-test-app captured on Pixel 6)_

https://user-images.githubusercontent.com/2678039/175370808-1a402d18-c004-4f4e-b029-edb091a6bff2.mp4

abhishek1508 commented on libnavui-resources/src/main/res/drawable/mapbox_bg_circle.xml:
```diff
@@ -1,5 +1,11 @@
-    <corners android:radius="56dp" />
-</shape>
\ No newline at end of file
+<ripple xmlns:android="http://schemas.android.com/apk/res/android"
```
This file should not be changed since it is used by stable buttons `MapboxAudioButton`, `MapboxRecenterButton` and `MapboxOverviewButton`. Changing this file results in the following:

#### Before
<img src="https://user-images.githubusercontent.com/9770186/175675827-061b5709-c676-44da-9983-fce954343fae.jpg" width="250" />

#### After
<img src="https://user-images.githubusercontent.com/9770186/175676333-75769021-4f8b-42e5-a2c3-db04b9561cc4.jpg" width="250" />

We should rather add a new circular drawable `xml` to `drop-in ui` module since it has a different design for how circular buttons should look like.

abhishek1508 commented on libnavui-resources/src/main/res/layout/extendable_button_layout.xml:
```diff
@@ -10,7 +10,6 @@
         android:id="@+id/buttonText"
         android:layout_width="@dimen/mapbox_button_size"
         android:layout_height="@dimen/mapbox_button_size"
-        android:background="@drawable/mapbox_bg_circle"
```
Let's not change this file.

abhishek1508 commented on libnavui-resources/src/main/res/layout/extendable_button_layout.xml:
```diff
@@ -28,7 +27,6 @@
         android:id="@+id/buttonIcon"
         android:layout_width="@dimen/mapbox_button_size"
         android:layout_height="@dimen/mapbox_button_size"
-        android:background="@drawable/mapbox_bg_circle"
```
Let's not change this file.

abhishek1508 commented on libnavui-voice/src/main/res/layout/mapbox_audio_guidance_button_layout.xml:
```diff
@@ -11,9 +11,11 @@
-        android:paddingStart="@dimen/mapbox_audioGuidanceButton_paddingStart">
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:paddingStart="10dp"
+        android:paddingTop="10dp"
+        android:paddingBottom="10dp">
```
```suggestion
        android:paddingStart="@dimen/mapbox_extendableButton_paddingStart"
        android:paddingTop="@dimen/mapbox_extendableButton_paddingStart"
        android:paddingBottom="@dimen/mapbox_extendableButton_paddingStart"
```

tomaszrybakiewicz replied:
The `MapboxExtendableButton` class and all its dimen (`mapbox_extendableButton_*`) and style values are in the `libnavui-dropin` module are not accessible from the `libnavui-voice`.

Should I move `MapboxExtendableButton` and `MapboxCameraModeButton` to the `libnavui-base`?

abhishek1508 reviewed (commented):
For the file `mapbox_audio_guidance_button_layout` the padding for `TextView` has a value of 8dp. It should be consistent with other extendable buttons.

```
<androidx.appcompat.widget.AppCompatTextView
            android:id="@+id/buttonText"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:paddingLeft="@dimen/mapbox_extendableButton_paddingStart". <-------
            android:text=""
            android:maxLines="1"
            android:textColor="@color/colorOnSurface"
            app:layout_constraintBottom_toBottomOf="@+id/iconImage"
            app:layout_constraintTop_toTopOf="@+id/iconImage"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toEndOf="@+id/iconImage" />
```

abhishek1508 commented:
@tomaszrybakiewicz 
When I running this branch locally and starting the `DropInButton` activity, it crashes for me

```
Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
        at android.view.LayoutInflater.createView(LayoutInflater.java:852)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:1004)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:959)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1131)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1092)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:1134)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:1092)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:680)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:532)
        at com.mapbox.navigation.qa_test_app.databinding.LayoutActivityDropinButtonsBinding.inflate(LayoutActivityDropinButtonsBinding.java:113)
        at com.mapbox.navigation.qa_test_app.databinding.LayoutActivityDropinButtonsBinding.inflate(LayoutActivityDropinButtonsBinding.java:107)
        at com.mapbox.navigation.qa_test_app.view.DropInButtonsActivity.onCreateContentView(DropInButtonsActivity.kt:27)
        at com.mapbox.navigation.qa_test_app.view.base.BaseNavigationActivity.initUi(BaseNavigationActivity.kt:333)
        at com.mapbox.navigation.qa_test_app.view.base.BaseNavigationActivity.init(BaseNavigationActivity.kt:273)
        at com.mapbox.navigation.qa_test_app.view.base.BaseNavigationActivity.onCreate(BaseNavigationActivity.kt:186)
        at com.mapbox.navigation.qa_test_app.view.DropInButtonsActivity.onCreate(DropInButtonsActivity.kt:32)
        at android.app.Activity.performCreate(Activity.java:8006)
        at android.app.Activity.performCreate(Activity.java:7990)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3595)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3786)
2022-06-24 14:32:24.005 26923-26923/com.mapbox.navigation.qa_test_app E/AndroidRuntime:     at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2257)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:233)
        at android.app.ActivityThread.main(ActivityThread.java:8030)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)
     Caused by: java.lang.ClassCastException: android.content.res.TypedArray cannot be cast to java.lang.AutoCloseable
        at com.mapbox.navigation.dropin.view.MapboxCameraModeButton.<init>(MapboxCameraModeButton.kt:99)
        at com.mapbox.navigation.dropin.view.MapboxCameraModeButton.<init>(MapboxCameraModeButton.kt:88)
        at com.mapbox.navigation.dropin.view.MapboxCameraModeButton.<init>(MapboxCameraModeButton.kt:82)
        	... 33 more
```

abhishek1508 commented on libnavui-dropin/src/main/res/layout/mapbox_camera_mode_button_layout.xml:
```diff
@@ -11,9 +11,11 @@
-        android:minWidth="72dp"
-        android:paddingStart="@dimen/mapbox_cameraModeButton_paddingStart">
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
```
I don't see the need for `android:layout_gravity="center"`. Work fine even after removing this line.

abhishek1508 replied:
I take this back

abhishek1508 commented on libnavui-dropin/src/main/res/layout/mapbox_extendable_button_layout.xml:
```diff
@@ -12,24 +12,25 @@
-        android:minHeight="@dimen/mapbox_extendableButton_minHeight"
-        android:minWidth="@dimen/mapbox_extendableButton_minWidth"
-        android:paddingStart="@dimen/mapbox_extendableButton_paddingStart">
+        android:layout_gravity="center"
```
I don't see the need for `android:layout_gravity="center"`. Work fine even after removing this line.

abhishek1508 commented on libnavui-voice/src/main/res/layout/mapbox_audio_guidance_button_layout.xml:
```diff
@@ -11,9 +11,11 @@
-        android:minWidth="72dp"
-        android:paddingStart="@dimen/mapbox_audioGuidanceButton_paddingStart">
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
```
I don't see the need for `android:layout_gravity="center"`. Work fine even after removing this line.

tomaszrybakiewicz commented:
> For the file mapbox_audio_guidance_button_layout, the padding for TextView has a value of 8dp. It should be consistent with other extendable buttons.

@abhishek1508. I am still working on this one. After revering back the padding to 8dp, I see buttons being asymmetrical when initially laid out in the shrunk state.

abhishek1508 commented:
> After revering back the padding to 8dp, I see buttons being asymmetrical when initially laid out in the shrunk state.

I am not sure I follow. Just so we are on the same page, `textView` `paddingLeft` is already set to 8dp for `MapboxAudioGuidanceButton` in `main` branch. My suggestion was to use `@dimen/mapbox_extendableButton_paddingStart`; the value of which is 10dp since both `MapboxCameraModeButton` and `MapboxExtendableButton` are using this res/dimens. `MapboxAudioGuidanceButton` should now be able to use `@dimen/mapbox_extendableButton_paddingStart` since we are moving it to `libnavui-resources`.

tomaszrybakiewicz commented:
> > After revering back the padding to 8dp, I see buttons being asymmetrical when initially laid out in the shrunk state.
> 
> I am not sure I follow. Just so we are on the same page, `textView` `paddingLeft` is already set to 8dp for `MapboxAudioGuidanceButton` in `main` branch. My suggestion was to use `@dimen/mapbox_extendableButton_paddingStart`; the value of which is 10dp since both `MapboxCameraModeButton` and `MapboxExtendableButton` are using this res/dimens. `MapboxAudioGuidanceButton` should now be able to use `@dimen/mapbox_extendableButton_paddingStart` since we are moving it to `libnavui-resources`.

The figma designs specify spacing between ImageView and the TextView as 8dp (which is what we used on main). In this PR I increased the spacing to 10dp to ensure that button looks symmetrical when shrunk. From your above comment, I understood that this change should be reverted back to 8dp. 

I think that asymmetry was always there, just not very visible due to the rectangular look. Now that the buttons can be square or circle, it is more prominent.

Anyway, I just pushed fix it in my latest commit. Now all buttons should look correct.

tomaszrybakiewicz commented:
@abhishek1508 I have moved `MapboxExtendableButton` to `libnavui-base` and `MapboxCameraModeButton` to `livnavui-maps` module. Updated both buttons to use the same dimen values and regenerated metalava files.
Please let me know if anything else needs to change.

abhishek1508 reviewed (approved):
Thanks for addressing the feedback. LGTM.
